File: src/main/java/dev/cdh/affiliate/CatWindow.java
package dev.cdh.affiliate;

import dev.cdh.constants.Behave;
import dev.cdh.constants.BubbleState;
import dev.cdh.constants.Layout;

import javax.swing.*;
import java.awt.*;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;

public final class CatWindow extends JWindow {
    private final Cat cat;

    public CatWindow(Cat cat) {
        this.cat = cat;
        setupWindow();
        setupMouseListeners();
        add(new Stage(cat));
    }

    private void setupWindow() {
        setType(Type.UTILITY);
        setSize(Layout.WINDOW_SIZE, Layout.WINDOW_SIZE);
        setPreferredSize(new Dimension(Layout.WINDOW_SIZE, Layout.WINDOW_SIZE));
        setLocationRelativeTo(null);
        setAlwaysOnTop(true);
        setBackground(new Color(0, 0, 0, 0));
    }

    private void setupMouseListeners() {
        MouseAdapter adapter = new MouseAdapter() {
            private final Point dragOffset = new Point(0, 0);

            @Override
            public void mousePressed(final MouseEvent e) {
                dragOffset.setLocation(e.getX(), e.getY());
                cat.stopWandering();
            }

            @Override
            public void mouseDragged(final MouseEvent e) {
                setLocation(e.getLocationOnScreen().x - dragOffset.x, e.getLocationOnScreen().y - dragOffset.y);
                if (cat.changeAction(Behave.RISING)) {
                    cat.animationState().resetFrame();
                }
            }

            @Override
            public void mouseReleased(final MouseEvent e) {
                if (cat.currentAction() == Behave.RISING) {
                    cat.changeAction(Behave.LAYING);
                    cat.animationState().resetFrame();
                }
            }

            @Override
            public void mouseClicked(final MouseEvent e) {
                cat.setBubbleState(BubbleState.HEART);
            }
        };
        addMouseListener(adapter);
        addMouseMotionListener(adapter);
    }
}

File: src/main/kotlin/affiliate/CatWindow.kt
package dev.cdh.affiliate

import dev.cdh.WINDOW_SIZE
import dev.cdh.constants.Behave
import dev.cdh.constants.BubbleState
import java.awt.Color
import java.awt.Dimension
import java.awt.Point
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import javax.swing.JWindow

class CatWindow(private val cat: Cat) : JWindow() {
    init {
        setupWindow()
        setupMouseListeners()
        add(Stage(cat))
    }

    private fun setupWindow() {
        setType(Type.UTILITY)
        setSize(WINDOW_SIZE, WINDOW_SIZE)
        setPreferredSize(Dimension(WINDOW_SIZE, WINDOW_SIZE))
        setLocationRelativeTo(null)
        setAlwaysOnTop(true)
        setBackground(Color(0, 0, 0, 0))
    }

    private fun setupMouseListeners() {
        val adapter: MouseAdapter = object : MouseAdapter() {
            private val dragOffset = Point(0, 0)

            override fun mousePressed(e: MouseEvent) {
                dragOffset.setLocation(e.getX(), e.getY())
            }

            override fun mouseDragged(e: MouseEvent) {
                setLocation(e.locationOnScreen.x - dragOffset.x, e.locationOnScreen.y - dragOffset.y)
                if (cat.changeAction(Behave.RISING)) {
                    cat.animationState.resetFrame()
                }
            }

            override fun mouseReleased(e: MouseEvent?) {
                if (cat.currentAction == Behave.RISING) {
                    cat.changeAction(Behave.LAYING)
                    cat.animationState.resetFrame()
                }
            }

            override fun mouseClicked(e: MouseEvent?) {
                cat.bubbleState = BubbleState.HEART
            }
        }
        addMouseListener(adapter)
        addMouseMotionListener(adapter)
    }
}